packages/apollo-server-core/src/requestPipeline.ts

```diff
@@ -162,35 +164,64 @@ export async function processGraphQLRequest<TContext>(
     requestContext,
   });
 
-  const parsingDidEnd = await dispatcher.invokeDidStartHook(
-    'parsingDidStart',
-    requestContext,
-  );
-
   try {
-    let document: DocumentNode;
-    try {
-      document = parse(query);
-      parsingDidEnd();
-    } catch (syntaxError) {
-      parsingDidEnd(syntaxError);
-      return sendErrorResponse(syntaxError, SyntaxError);
+    // If we're configured with a document store, we'll utilize the hash of the
+    // operation in order to lookup a previously parsed-and-validated operation
+    // from that.  A failure to retrieve anything from the cache simply means
+    // we're guaranteed to do the parsing and validation ourselves.
+    let document: DocumentNode | undefined;
+
+    if (config.documentStore) {
+      document = await config.documentStore.get(queryHash);
     }
 
-    requestContext.document = document;
+    // If we still don't have a document, we'll need to parse and validate it.
+    // With success, we'll attempt to save it into the store for future use.
+    if (!document) {
+      const parsingDidEnd = await dispatcher.invokeDidStartHook(
+        'parsingDidStart',
+        requestContext,
+      );
 
-    const validationDidEnd = await dispatcher.invokeDidStartHook(
-      'validationDidStart',
-      requestContext as WithRequired<typeof requestContext, 'document'>,
-    );
+      try {
+        requestContext.document = document = parse(query);
+        parsingDidEnd();
+      } catch (syntaxError) {
+        parsingDidEnd(syntaxError);
+        return sendErrorResponse(syntaxError, SyntaxError);
+      }
+
+      const validationDidEnd = await dispatcher.invokeDidStartHook(
+        'validationDidStart',
+        requestContext as WithRequired<typeof requestContext, 'document'>,
+      );
 
-    const validationErrors = validate(document);
+      const validationErrors = validate(document);
 
-    if (validationErrors.length === 0) {
-      validationDidEnd();
-    } else {
-      validationDidEnd(validationErrors);
-      return sendErrorResponse(validationErrors, ValidationError);
+      if (validationErrors.length === 0) {
+        validationDidEnd();
+      } else {
+        validationDidEnd(validationErrors);
+        return sendErrorResponse(validationErrors, ValidationError);
+      }
+
+      if (config.documentStore) {
+        // The underlying cache store behind the `documentStore` returns a
+        // `Promise` which is resolved (or rejected), eventually, based on the
+        // success or failure (respectively) of the cache save attempt.  While
+        // it's certainly possible to `await` this `Promise`, we don't care about
+        // whether or not its successful at this point.  We'll instead proceed
```

Comment:
"it is" or "it's"

Reply:
Thanks.  45f14a297151129ead30cf68e93b64a777cf02ce.